### ISSUE TITLE
Technical analysis: re-running Wikipedia exports

### ISSUE THREAD

**Issue by MichaelKohler** [common-voice/cv-sentence-extractor#147]
For several languages we've run a Wikipedia extract in the past. Since these exports new Wikipedia articles have been created. If legally possible, we'd like to re-run the export on these new articles. The legal question is out of scope for this issue here, we can clear that once we know that it's technically possible.

**Comment by MichaelKohler** [comment]
Here's how a random page looks like in the wikipedia dump:

```
<page>
    <title>M<C3><BC>lhausen</title>
    <ns>0</ns>
    <id>1632</id>
    <revision>
      <id>994091</id>
      <parentid>975031</parentid>
      <timestamp>2021-05-10T04:33:27Z</timestamp>
      <contributor>
        <username>Holder</username>
        <id>491</id>
      </contributor>
      <minor />
      <model>wikitext</model>
      <format>text/x-wiki</format>
      <text bytes="18642" xml:space="preserve">{{DISPLAYTITLE:Milh<C3><BC>sa}}
{{Infobox Gemeinde in Frankreich
| nomcommune = Mulhouse
| nomalsacien = Milh<C3><BC>sa
| alt moy = 240
| alt mini = 232
| alt maxi = 338
| km<C2><B2> = 22.18
}}
............. more from the content of the page ..............
      <sha1>ml0ydbtc3dvlaagdv2kr953eqq99hmh</sha1>
    </revision>
  </page>
```

Note that the `<timestamp>` is not when the page was created, but when it was last updated. Also the `<id>` and `<parentid>` do not help us here.

**Comment by MichaelKohler** [comment]
https://en.wikipedia.org/wiki/Special:Export would need a lot of work, and as far as I see it still does not include the full revision history. The article I tried didn't have more than 1000 revisions (limit) but still didn't include the first revision.

**Comment by MichaelKohler** [comment]
The first revision date for any page can be found with the following request: `https://en.wikipedia.org/w/api.php?action=query&prop=revisions&rvlimit=1&rvprop=timestamp&rvdir=newer&titles=<insertPageTitleHere>`. However I don't think this scales well...

**Comment by MichaelKohler** [comment]
Another option would have been to download the dump from when the extract was done, and then compare 1:1 with the current dump to find new pages added since then. However it seems that https://dumps.wikimedia.org/dewiki/ does not go that far back.

**Comment by MichaelKohler** [comment]
The `dewiki-latest-pages-meta-history*.xml*` files probably contain the full history of all articles. The article I checked indeed had revisions all the way back to 2004.

I'm not sure what files we'd need exactly, would need to figure out what the `p109531p1` means in the file `dewiki-latest-pages-meta-history1.xml-p109531p1` for example. There are multiple `history1.xml` all with a different id. We might need just one of those, or all of them. This might very well be "part" and we need all.

If that's the case, we'd need to download roughly 6GB for `history1` and there are 6 history sections in total. Didn't calculate the rest, but assuming roughly same file sizes that would amount to 36GB. With the same compression ratio as I had in the file I downloaded, this would amount to roughly 850 GB of data for one language. While these files could be processed one after the other we wouldn't need as much storage space, but I'd count on having at least 2 part files at the same time so we can run a script while another part is downloading.

I have absolutely no good guess right now how long it would take to analyze all these files and get the creation date for all the articles.

**Comment by MichaelKohler** [comment]
https://wikimedia.org/api/rest_v1/#/Edited%20pages%20data seems to provide a *count* of new pages. For example https://wikimedia.org/api/rest_v1/metrics/edited-pages/new/en.wikipedia/all-editor-types/content/monthly/20210101/20210626 lists the count of newly created pages in the English wikipedia by month since the beginning of this year.

Looking at different APIs next, maybe there might be something.

**Comment by MichaelKohler** [comment]
There we go, what we want is probably https://en.wikipedia.org/w/api.php?action=query&list=recentchanges&rctype=new&rclimit=500&rcnamespace=0&rcstart=now&rcend=1609455600&format=json which would list the new page changes since the start of this year. However, given how many articles are created daily and the request limit is 500, this still would end up being quite a lot of requests.

**Comment by Mte90** [comment]
I used the query wikidata system to get the latest new wikisource published but as origin after a specific date for italian, so we can use to see if we need to update the stuff for the deepspeech model

https://query.wikidata.org/#SELECT%20DISTINCT%20%3Fitem%20%3FwikisourceSitelink%20WHERE%20%7B%0A%20%20%3FwikisourceSitelink%20schema%3AisPartOf%20%5B%20wikibase%3AwikiGroup%20%22wikisource%22%20%5D%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20schema%3AinLanguage%20%27it%27%3B%0A%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20%20schema%3Aabout%20%3Fitem.%0A%20%20%3Fitem%20wdt%3AP577%20%3Fdate.%0A%20%20FILTER%28%3Fdate%20%3E%20%221914-01-01T00%3A00%3A00Z%22%5E%5Exsd%3AdateTime%29%0A%7D%0AORDER%20BY%20DESC%20%28%3Fitem%29%0ALIMIT%20100

Probably we can see if we can do it for wikipedia.

My question about the whole idea is that we are just getting new pages but not getting sentences for updated pages (I guess that is more complicated for them).

**Comment by Heyhillary** [comment]
Hoya, just to follow up.

Legal is happy with us proceeding as long as the process is the same or a re-run.

Thanks for your patience.

**Comment by MichaelKohler** [comment]
> Probably we can see if we can do it for wikipedia.

So far I couldn't find the same interface for Wikipedia. Did you have more luck there?

There is the possibility to request a query on the database though, this might be easier than using the API. https://en.wikipedia.org/wiki/Wikipedia:Request_a_query

> My question about the whole idea is that we are just getting new pages but not getting sentences for updated pages (I guess that is more complicated for them).

Yes, we only want new articles, we can't query existing articles, even if they were updated.

**Comment by Mte90** [comment]
Yes you can do a query, I am already doing it one with the link shared in the other comment.
When I talked to the wikipedia community they suggested to use that tool

**Comment by MichaelKohler** [comment]
@Mte90 FYI, so far I haven't found a query that actually would do what we want to do on Wikidata.

**Comment by MichaelKohler** [comment]
> There we go, what we want is probably https://en.wikipedia.org/w/api.php?action=query&list=recentchanges&rctype=new&rclimit=500&rcnamespace=0&rcstart=now&rcend=1609455600&format=json which would list the new page changes since the start of this year.

Unfortunately this isn't true, I should have actually tested this. This only gets the last 30 days, no matter if the `rcend` param is further in the past. :(

**Comment by Mte90** [comment]
So maybe we can add a monthly check?

**Comment by MichaelKohler** [comment]
I really hope that's not the only way :(
